File: java/BubbleSort.java
public class BubbleSort {
    public <T extends Comparable<T>> void sort(T[] arr) {
        int n = arr.length - 1;
        for (int i = 0; i < n; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - i; j++) {
                if (arr[j].compareTo(arr[j + 1]) > 0) {
                    T temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] =  temp;
                    swapped = true;
                }
            }

            if (swapped == false)
                break;
        }
    }
}

File: src/bubble_sort.rs
pub enum SortOrder {
    Asc,
    Desc,
}
pub fn bubble_sort<T: PartialOrd>(arr: &mut [T], ord: SortOrder) {
    let n = arr.len() - 1;
    for i in 0..n {
        let mut swapped = false;
        for j in 0..n - i {
            match ord {
                SortOrder::Asc => {
                    if arr[j] > arr[j + 1] {
                        arr.swap(j, j + 1);
                        swapped = true;
                    }
                },
                SortOrder::Desc => {
                    if arr[j] < arr[j + 1] {
                        arr.swap(j, j + 1);
                        swapped = true;
                    }
                },
            }
            
        }
        if !swapped {
            break;
        }
    }
}